File: languages/typescript/03-advanced/constructors-and-invariants/example/main.ts
// Module focus: Building objects that start valid and stay valid through guarded updates.
// Why it matters: practicing constructors and invariants patterns makes exercises and checkpoints easier to reason about.

class BankAccount {
    #balance: number;

    constructor(
        readonly owner: string,
        openingBalance: number,
    ) {
        // Reject invalid construction so no BankAccount starts in a broken state.
        if (!owner.trim()) {
            throw new Error("Owner name is required.");
        }
        if (openingBalance < 0) {
            throw new Error("Opening balance must be non-negative.");
        }
        this.#balance = openingBalance;
    }

    deposit(amount: number): void {
        // Guard every state-changing method, not only the constructor.
        if (amount <= 0) {
            throw new Error("Deposit amount must be positive.");
        }
        this.#balance += amount;
    }

    withdraw(amount: number): boolean {
        if (amount <= 0 || amount > this.#balance) {
            return false;
        }
        this.#balance -= amount;
        return true;
    }

    printStatus(): void {
        // Report output values so learners can verify the constructors and invariants result.
        console.log(`${this.owner}: ${this.#balance.toFixed(2)}`);
    }
}

// The valid scenario shows normal updates after the invariant checks pass.
const account = new BankAccount("Ana", 120);
account.deposit(30);
account.withdraw(50);
account.printStatus();

export {};

File: languages/java/03-advanced/constructors-and-invariants/example/Main.java
// Module focus: Building objects that start valid and stay valid through guarded updates.
// Why it matters: practicing constructors and invariants patterns makes exercises and checkpoints easier to reason about.

import java.util.Locale;

public class Main {
    static class Temperature {
        private double celsius;

        Temperature(double celsiusValue) {
            // The constructor clamps invalid starting data so every object starts valid.
            celsius = Math.max(-273.15, celsiusValue);
        }

        boolean setCelsius(double newValue) {
            // Updates must protect the same invariant as construction.
            if (newValue < -273.15) {
                return false;
            }

            celsius = newValue;
            return true;
        }

        double celsius() {
            // Accessors expose the valid state without letting callers edit fields directly.
            return celsius;
        }
    }

    public static void main(String[] args) {
        Locale.setDefault(Locale.US);

        // Prepare sample inputs that exercise the key constructors and invariants path.
        Temperature temperature = new Temperature(-500.0);

        // Report values so learners can verify the constructors and invariants outcome.
        System.out.printf("Initial value (clamped): %.2f C%n", temperature.celsius());

        boolean updated = temperature.setCelsius(25.0);
        System.out.println("Set to 25.0 success: " + updated);
        System.out.printf("Current value: %.2f C%n", temperature.celsius());

        boolean rejected = temperature.setCelsius(-300.0);
        System.out.println("Set to -300.0 success: " + rejected);

        System.out.printf("Current value: %.2f C%n", temperature.celsius());
    }
}
